src/main/scala/com/fulcrumgenomics/util/miseq/SampleSheet.scala

```diff
@@ -0,0 +1,152 @@
+  */
+package com.fulcrumgenomics.util.miseq
+
+import java.nio.file.Path
+
+import scala.collection.mutable.ListBuffer
+import scala.io.Source
+
+/**
+  * Stores information about samples from an Illumina Experiment Manager Sample Sheet (typically a MiSeq).  The
+  * samples may also include derived information.  If library id is not specified, it will be set to the sample name.
+  *
+  * Optional fields include support for specifying the expected sample barcode for each sample.  The sample barcode
+  * can be present as a sub-sequence (or sub-sequences) in the i7 or i5 read.  If additional bases are found in the i7
+  * or i5 read, such as molecular barcodes, the should be included as Ns.  It is up to the developer to obtain the
+  * correct read structure elsewhere to infer which bases are sample barcode and which bases are not (ex. molecular
+  * identifiers).  If the sample barcode is inline in either read of a pair, the sample barcode sequence can be specified
+  * here.  In this case, only include the sample barcode, and not the additional bases, such as template bases.
+  *
+  * NB: the lookup in the columns in the Sample Sheet is case insensitive.
+  *
+  * @author Nils Homer
+  */
+object SampleSheet {
+  /** Required header names for standard fields in the Illumina Experiment Manager Sample Sheet. */
+  val SampleId: String       = "Sample_ID"
+  val SampleName: String     = "Sample_Name"
+  val LibraryId: String      = "Library_ID"
+  val SampleProject: String  = "Sample_Project"
+  val Description: String    = "Description"
+  /** Optional header names for standard fields in the Illumina Experiment Manager Sample Sheet. */
+  val Lane: String           = "Lane"
+  val R1BarcodeBases: String = "R1_Barcode_Bases"
+  val R2BarcodeBases: String = "R2_Barcode_Bases"
+  val I7Bases: String        = "Index"
+  val I5Bases: String        = "Index2"
+
+  /** Attempts to clean special characters from a sample id or name as Illumina's software does. */
+  def cleanMiseqSampleId(id: String): String = {
+    id.map {
+      case '#' => ' '
+      case '_' | '+' | ' ' => '-'
+      case c => c
+    }
+  }
+
+  /** Gets the trimmed value of the given key (`name`) from the map.  If empty or not found, returns None */
+  private def getStringField(sampleDatum: Map[String, String], name: String): Option[String] = {
+    sampleDatum.get(name.toUpperCase) match {
+      case None => None
+      case Some(str) if str.isEmpty => None
+      case Some(str) => Some(str.trim)
+    }
+  }
+
+  private def getIntField(sampleDatum: Map[String, String], name: String): Option[Int] = {
+    getStringField(sampleDatum, name).map(_.toInt)
+  }
+
+  private[miseq] val SplitRegex = ",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)"
+}
+
+/** Create a sample sheet from the given file (typically 'SampleSheet.csv'). If the library id is omitted, the
+  * sample name will be used in its stead. */
+class SampleSheet(sampleSheet: Path, lane: Option[Int] = None) extends Iterable[Sample] {
```

Comment:
Looking at this again, it feels like it would be nicer to have the primary constructor take the set of samples in, and have this code in a `SampleSheet.apply` method - would make testing much simpler.